src/qutip_qip/decompose/decompose_two_qubit_gate.py

```diff
@@ -0,0 +1,298 @@
+        phase_gate,
+    )
+
+
+def _decompose_U5(input_array):
+    """Decomposes a two-level unitary into ABC and CNOT when target = 0 and
+    control value is 1.
+
+    Based on Lemma 5.1 of arXiv:quant-ph/9503016v1 with flipped targets.
+    """
+    U = np.array(
+        [[input_array[1][1], input_array[1][3]],
+            [input_array[3][1], input_array[3][3]]]
+    )
+    check_gate(Qobj(U), num_qubits=1)
+    gate_list = decompose_one_qubit_gate(Qobj(U), "ZYZ_PauliX")
+
+    # No need to change targets of A, B, C - default is already 0
+    # last is phase whose target is changed based on Lemma 5.2 of
+    # arXiv:quant-ph/9503016v1 alongwith the gate type
+    phase_angle = gate_list[-1].arg_value
+    phase_gate = Gate(
+        "PHASEGATE",
+        targets=[1],
+        arg_value=phase_angle,
+        arg_label=r"{:0.2f} \times \pi".format(phase_angle / np.pi),
+    )
+    Pauli_X = Gate("X", targets=[0], classical_controls=[0])
+
+    CNOT_ctrl_1 = Gate("CNOT", controls=1, targets=0)
+    return (
+        gate_list[0],
+        gate_list[1],
+        Pauli_X,
+        CNOT_ctrl_1,
+        Pauli_X,
+        gate_list[3],
+        gate_list[4],
+        Pauli_X,
+        CNOT_ctrl_1,
+        Pauli_X,
+        gate_list[6],
+        phase_gate,
+    )
+
+
+def _decompose_to_two_level_arrays(input_gate):
+    """Output of two qubit gate in terms of two-level numpy arrays.
+    """
+    check_gate(input_gate, num_qubits=2)
+    input_array = input_gate.full()
+
+    # Calculate the two level numpy arrays
+    array_list = []
+    index_list = [[0, 1], [0, 2], [0, 3], [1, 2], [1, 3]]
+    for i in range(len(index_list)):
+        index_1, index_2 = index_list[i]
+
+        # Values of single qubit U forming the two level unitary
+        a = input_array[index_1][index_1]
+        a_star = np.conj(a)
+        b = input_array[index_2][index_1]
+        b_star = np.conj(b)
+        norm_constant = cmath.sqrt(np.absolute(a*a_star)+np.absolute(b*b_star))
+
+        # Create identity array and then replace with above values for index_1
+        # and index_2
+        U_two_level = np.identity(4, dtype=complex)
+        U_two_level[index_1][index_1] = a_star/norm_constant
+        U_two_level[index_2][index_1] = b/norm_constant
+        U_two_level[index_1][index_2] = b_star/norm_constant
+        U_two_level[index_2][index_2] = -a/norm_constant
+
+        # Change input by multiplying by above two-level
+        input_array = np.dot(U_two_level, input_array)
+
+        # U dagger to calculate the gates
+        U__two_level_dagger = np.transpose(np.conjugate(U_two_level))
+        U__two_level_dagger = Qobj(U__two_level_dagger, dims=[[2, 2], [2, 2]])
+        array_list.append(U__two_level_dagger)
+
+    # for U6 - multiply input array by U5 and take dagger
+    U6_dagger = input_array
+    array_list.append(Qobj(U6_dagger, dims=[[2, 2], [2, 2]]))
+    return(array_list)
+
+
+def decompose_two_qubit_to_two_level_unitary(input_gate):
```

Comment:
Same as above. Once a general method for decomposition to two level unitaries exists, this function will be discarded. 

Defined here for convenience of checking the output.